src/engine.rs

```diff
@@ -251,37 +251,77 @@ impl Engine {
         origin: Address,
         value: U256,
         input: &[u8],
-    ) -> (ExitReason, Address) {
+    ) -> (ExitReason, SubmitResult) {
         let mut executor = self.make_executor();
         let address = executor.create_address(CreateScheme::Legacy { caller: origin });
         let (status, result) = (
             executor.transact_create(origin, value, Vec::from(input), u64::MAX),
             address,
         );
+        let used_gas = executor.used_gas();
         let (values, logs) = executor.into_state().deconstruct();
-        self.apply(values, logs, true);
-        (status, result)
+        self.apply(values, Vec::<Log>::new(), true);
+
+        let mut res_logs = Vec::new();
+        for log in logs {
+            res_logs.push(log.into());
+        }
+
+        let (res_status, used_gas) = match status {
+            ExitReason::Succeed(_) => (true, used_gas),
+            ExitReason::Revert(_) => (false, used_gas),
+            ExitReason::Error(_) | ExitReason::Fatal(_) => (false, 0),
```

Comment:
Is it true that the gas used is 0 in the case of `Error`? That doesn't seem right to me. For example in the case of an out-of-gas error all gas was spent.

Reply:
The `ExitReason::Error` and `ExitReason::Fatal` are non-opcode related, and the `used_gas` is returned to signify how much gas was spent as for both `Succeed` and `Revert` it is consumed.

Reply:
Ok, so `executor.used_gas()` may not be zero (and probably isn't I think) in the `Error` case, but we are choosing to have the semantics that `SubmitResult.gas_used` is always zero in the case of an error. Could you elaborate more on why this is a good semantic to have?

Assuming this is a property we want to enforce then I think we should do it at the type level. I like the guideline of trying to "make invalid states unrepresentable". For example here it should be impossible for `SubmitResult.gas_used > 0` and `SubmitResult.status == false`, but that is not the case as the struct exists currently. Maybe a better representation would be to drop the `status` field and have `gas_used: Option<u64>`, where the `None` case would then represent error. I'm open to other ways of factoring this as well if you think the `Option` is a little awkward, for example `Result<u64, ErrorSingleton>`.

Reply:
I do agree with that philosophy. So, looking at the code of the EVM, there would be a return value on either `Succeed` or `Revert`. `Fatal` and `Error` return a `Vec::new()`. So, it could just simply return `gas_used: Option<u64>` and `result: Option<Vec<u8>>`. But, if they both are going to be optional fields, I tend to like to group those together, perhaps as `SubmitReturn`. Then, you would either have that, or not. I think I'll do that, given what I saw with the fact that `result` too would be returning an empty `Vec` per inner operations.

Reply:
As I recall previous discussion on this, the idea was that in the third case we've hit an error not in the user contract being executed, but in the Aurora Engine itself. In that case, we shouldn't charge EVM gas. However, is it really the case that only internal errors will end up in the third case here and everything else will cleanly be caught and returned as a revert?

Reply:
@joshuajbouw Thoughts? Still planning on doing something with regards to this?

Reply:
Looked quite deeply into this. It looks like we are required to handle all these ourselves. The disappointing thing is that all the returns are not organised. I'm going to have to submit a PR urgently that will handle which are failures that cost gas, and which do not.